### DISCUSSION TITLE
Removal of the device from Zigpy

### DISCUSSION THREAD

**Discussion by Adminiuga** [zigpy/zigpy#572] [RFC]
This discussion is based on https://github.com/zigpy/zigpy/pull/564#issuecomment-737320694

1. `ControllerApplication.force_remove()` method stems from the original bellows implementation and was meant to call `removeDevice()` EZSP command. Other radio libs don't have similar commands and often just "plug" this method.
2. ZHA removes the device from HA, regardless if the `ControllerApplication.remove()` method worked or not. And with calling of the `force_remove()` method it is implied that the device was removed gracefully or forcibly
3. NCP images 6.6 and later have an unexpected surprise: sending a ZDO leave message to a directly connected "child" results in not receiving a `messageSentHandler`.  `bellows` expects a messageSentHandler for every `sendUnicast` or `sendMulticast` and missing message results in an extensive timeout.

With the above, what if for `ControllerApplication.remove()` method we do the following:

1. schedule a catching coro task for `device_to_remove.zdo.leave()`
2. if the `device_to_remove` is an end device, then schedule a catching coro task of sending a ZDO leave message to each device which has it in the neighbour tables. In other words, send a ZDO leave with IEEE of a children to all active and old parents of the device
3. remove the device from the ControllerApplication.devices dict. Or maybe schedule it for removal in 30s, to allow any replies to be properly processed.?

@puddly @dmulcahey

**Comment by puddly** [comment]
As long as ZHA can deal with the potential 30s timeout in case the device is offline, I think notifying all potential parents for end devices is a good idea.

As for delaying removal, would you be getting a reply from the device if you cannot contact it to begin with with the direct ZDO request?

**Comment by Adminiuga** [comment]
> As for delaying removal, would you be getting a reply from the device if you cannot contact it to begin with with

No, the delayed removal is for the online devices for which we will be getting a response.  Could just put into a separate coro and after a 30s pop it from the devices, after the request to the `device_to_remove` was sent.

**Comment by Hedda** [comment]
> 2\. ZHA removes the device from HA, regardless if the `ControllerApplication.remove()` method worked or not. And with calling of the `force_remove()` method it is implied that the device was removed gracefully or forcibly

Someone by the way requested here that use of ZHA.REMOVE be documented for HA's ZHA -> https://github.com/home-assistant/home-assistant.io/issues/17072

**Reply by Adminiuga** [reply]
It's documented. The service call from the developer tab does list all the parameters
